### PR TITLE
Connect NMEA output to ntpd via pty linked at /dev/gps0

### PR DESCRIPTION
**Pull request by gsmcmullin**
This provides UTC from our navigation solution as a reference clock to set the Linux system time.
- Start a new zmq_adapter to connect NMEA stream and use socat to link to pty at /dev/gps0
- Add NMEA reference clock to ntpd configuration
- Patched NTP to allow a system time older than it's build time in NMEA driver
- Add a setting to select system time source for cases when NTP and GPS time are available, but different.
  
For https://github.com/swift-nav/firmware_team_planning/issues/290

### CODE REVIEW

**Comment by gsmcmullin** [comment]
ntpd, as provided by buildroot has a problem if the time provided by the GPS is older than the package build time.  In real life situations this isn't a problem, but it will cause problems when connecting to a simulator or pre-recorded RF.

**Comment by gsmcmullin** [comment]
For future expansion:
- Add an SBP reference clock driver to ntpd
- Add PPS synchronization
- Bring in GPSD

**Comment by anonymous-travis-swiftnav** [comment]
## v1.2.0-develop-2017110106-2-gcc7a0cc
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.2.0-develop-2017110106-2-gcc7a0cc/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.2.0-develop-2017110106-2-gcc7a0cc/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.2.0-develop-2017110106-2-gcc7a0cc

**Comment by gsmcmullin** [comment]
Experiments to reconfigure ntpd at runtime were unsuccessful.  Although it is possible to update the configuration using ntpq or ntpdc, I was unable to get it to reliably change the clock source.
The setting introduced is a sensible default, but allows the user to explicitly select NTP or GPS as a time source, when, for example, connected to a simulator where GPS time and NTP time may both be available, but provide significantly different times.

**Comment by anonymous-travis-swiftnav** [comment]
## v1.2.0-develop-2017122916-5-g1ab12ce
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.2.0-develop-2017122916-5-g1ab12ce/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.2.0-develop-2017122916-5-g1ab12ce/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.2.0-develop-2017122916-5-g1ab12ce

File: board/piksiv3/rootfs-overlay/etc/ntp.conf.gps [hunk 1]
```diff
@@ -0,0 +1,7 @@
+server 127.127.20.0 mode 1 true
```
Comment (silverjam) [inline board/piksiv3/rootfs-overlay/etc/ntp.conf.gps]: Where does this value come from? `127.127.20.0`?

File: board/piksiv3/rootfs-overlay/etc/init.d/S45zmq_adapter_gps0 [hunk 2]
```diff
@@ -0,0 +1,88 @@
+#!/bin/sh
+
+name="zmq_adapter_gps0"
+cmd="zmq_adapter --stdio -s '>tcp://127.0.0.1:44030' | socat pty,raw,link=/dev/gps0 fd:0"
```
Comment (silverjam) [inline board/piksiv3/rootfs-overlay/etc/init.d/S45zmq_adapter_gps0]: I think later we can consolidate this into the new nmea_daemon.

File: board/piksiv3/rootfs-overlay/etc/ntp.conf.gps (hunk 1 above)
Comment (silverjam) [reply board/piksiv3/rootfs-overlay/etc/ntp.conf.gps]: I guess this is a standard ntpd thing, it looks like it reads from /dev/gps0 and add a "server" for gps at 127.127.20.0?  Would appreciate a comment to that effect for the uninitiated.

http://support.ntp.org/bin/view/Support/ConfiguringNMEARefclocks
http://wiki.gude.info/Ntpd_Installation

**Comment by anonymous-travis-swiftnav** [comment]
## v1.2.0-develop-2018010520-10-gbf86a24
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.2.0-develop-2018010520-10-gbf86a24/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.2.0-develop-2018010520-10-gbf86a24/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.2.0-develop-2018010520-10-gbf86a24

**Comment by anonymous-travis-swiftnav** [comment]
## v1.2.0-develop-2018010520-10-g68759a0
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.2.0-develop-2018010520-10-g68759a0/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.2.0-develop-2018010520-10-g68759a0/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.2.0-develop-2018010520-10-g68759a0

File: board/piksiv3/rootfs-overlay/etc/init.d/S45zmq_adapter_gps0 (hunk 2 above)
Comment (gsmcmullin) [reply board/piksiv3/rootfs-overlay/etc/init.d/S45zmq_adapter_gps0]: I agree

**Pull request merged**